cpp/tests/trajectory/test_trajectory_distance_speed.cu

```diff
@@ -0,0 +1,206 @@
+                   [](int32_t i) { return i * ms_per_hour + i; }); 
+
+    std::vector<double> x(sequence.begin(), sequence.end());
+    wrapper<double> in_x{x};
+        //[&](gdf_index_type i) { return static_cast<double>(sequence[i]); });
+    wrapper<double> in_y(column_size,
+        [&](gdf_index_type i) { return static_cast<double>(sequence[i]); });
+    wrapper<cudf::timestamp> in_ts(column_size,
+        [&](gdf_index_type i) {
+            return static_cast<cudf::timestamp>(ms_vector[sequence[i]]);
+        });
+
+    std::vector<gdf_size_type> trajectory_length{2 * column_size / 3, 
+                                                 (column_size + 5) / 6,
+                                                 (column_size + 5) / 6};
+    std::vector<gdf_size_type> trajectory_offset{2 * column_size / 3,
+                                                 5 * column_size / 6,
+                                                 column_size};
+    wrapper<gdf_size_type> traj_len(trajectory_length);
+    wrapper<gdf_size_type> traj_offset{trajectory_offset};
+
+    std::pair<gdf_column, gdf_column> distance_speed;
+
+    EXPECT_NO_THROW(
+     distance_speed = cuspatial::trajectory_distance_and_speed(in_x, in_y,
+                                                               in_ts, traj_len,
+                                                               traj_offset);
+    );
+
+    EXPECT_EQ(distance_speed.first.size, 3);
+    EXPECT_EQ(distance_speed.second.size, 3);
+
+    std::vector<gdf_size_type> id{0, 1, 2};
+    std::vector<double> distance(3, 0);
+    std::vector<double> time(3, 0);
+    std::vector<double> speed(3, 0);
+    
+    // compute expected distance and speed
+    for (auto i : id) {
+        gdf_size_type length = trajectory_length[i];
+        gdf_size_type begin = (i == 0) ? 0 : trajectory_offset[i - 1];
+        gdf_size_type end = trajectory_offset[i];
+        EXPECT_EQ(end, begin + length);
+
+        for (gdf_size_type j = 0; j < length-1; j++) {
+            double d = static_cast<double>(sequence[begin + j + 1] -
+                                           sequence[begin + j]);
+            double ms = static_cast<double>(ms_vector[begin + j + 1] -
+                                            ms_vector[begin + j]);
+            d = sqrt(2 * d * d); // since x == y
+            distance[i] += d * m_per_km;
+            time[i] += (ms / ms_per_second);
+        }
+        speed[i] = distance[i] / time[i];
+    }
+
+    std::vector<double> gpu_distance(distance_speed.first.size);
+    cudaMemcpy(gpu_distance.data(), distance_speed.first.data,
+               distance_speed.first.size * sizeof(double), cudaMemcpyDefault);
+    std::vector<double> gpu_speed(distance_speed.second.size);
+    cudaMemcpy(gpu_speed.data(), distance_speed.second.data,
+               distance_speed.second.size * sizeof(double), cudaMemcpyDefault);
+    EXPECT_THAT(gpu_distance, testing::Pointwise(FloatNearPointwise(1e-8),
```

Comment:
Can this just use GTests `EXPECT_FLOAT_EQ` and `EXPECT_DOUBLE_EQ` instead? These macros compare based on units of least precision rather than a fixed tolerance.

https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#floating-point-comparison

Reply:
I hadn't seen that. Why do they have matchers in both gtest and gmock? Anyway, I couldn't get `testing::Eq` to instantiate without an argument, which means it couldn't be used for a container. This point wise thing was all I could find that would compile.

Reply:
The difference between CPU and GPU results is in the very small bits in this case...

Reply:
I looked at that. EXPECT_FLOAT_EQ only compares two values, not two containers.

Reply:
You're using a tool from Gmock meant for making mock classes for testing class interfaces.

For asserting two vectors are equal, I'd iterate over both and use `EXPECT_FLOAT_EQ`.

Reply:
https://github.com/google/googletest/blob/master/googlemock/docs/cook_book.md#using-matchers-in-googletest-assertions

> Using Matchers in googletest Assertions
> 
> Since matchers are basically predicates that also know how to describe themselves, there is a way to take advantage of them in googletest assertions. It's called ASSERT_THAT and EXPECT_THAT:
> 
> ASSERT_THAT(value, matcher);  // Asserts that value matches matcher.
> EXPECT_THAT(value, matcher);  // The non-fatal version.

Reply:
gmock lets me compare two std::vectors without reinventing the wheel, and it prints useful error messages when they don't match.